float64[] state1
float64[] state2
float64[] joint
int32 type
